teps/XXXX-task-execution-status-at-runtime.md

```diff
@@ -0,0 +1,115 @@
+---
+title: task-exec-status-at-runtime
+authors:
+  - "@pritidesai"
+creation-date: 2020-10-15
+last-updated: 2020-10-19
+status: proposed
+---
+
+# TEP-0027: Task Execution Status at Runtime
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+**Note** In this TEP, `dag` tasks refer to `pipelineTasks` under `tasks` section and `finally` tasks refer to `pipelineTasks` in `finally` section.
+
+A `finally` task in a `pipeline` is executed just before exiting the `pipeline`. `Pipeline` executes `finally` tasks
+after all `dag` tasks finish successfully or one of the `dag` task results in failure i.e `finally` tasks are always
+executed in parallel irrespective of success or failure of a `dag` task. Such `finally` tasks are lacking access to the
+execution status of the `dag` tasks at runtime to better design and process success scenarios v/s failure scenarios.
+
+This proposal is enabling a `finally` task to access the execution status of `dag` tasks at runtime to better design
+`pipeline` based on the status.
+
+## Motivation
+
+* A `pipeline` author wants to send notification to an external system based on the execution status of the `dag` tasks. 
+* A `pipeline` author wants to rollback if the deployment fails in that environment/cluster/region/provider.
+
+### Goals
+
+The main goal of this proposal is to allow access to the execution status of a task within the `pipeline` at runtime.
+And design this approach in the simplest possible way, without any extra processing. 
+
+An additional but important goal of this proposal is identify the different execution states of a `pipelineTask` and
+how each of those states derived based on `taskRun` and/or `pipelineRun` status.
+
+### Non-Goals
+
+* This proposal does not allow dag tasks to access status of other dag tasks.
+
+* This proposal is not to expose the execution state of `finally` tasks since its not required.
+ 
+* This proposal is not to expose the entire status of a `taskRun`/`pipelineRun` as metadata at runtime.
+
+## Proposal
+
+Introduce a new context variable `$(context.pipelineRun.Tasks.<pipelineTask>.status)` which resolves to one of the execution states:
```

Comment:
I want to pick up the thread that @sbwsg started - I can't seem to reply to those comments so apologies I have to start again XD

having both `tasks.<taskName>.results.<resultName>` and `context.` with task execution info does feel a little odd. 

@pritidesai mentioned:

> Status is something not generated by the task itself, its something controller generating for each task which I think falls under context.pipelineRun similar to pipelineRun name, id, etc.

So it seems like we're drawing a line around `context` that has to do with values the controller is generating - however most of the values we have in there right now (name, namespace) are set outside of the controller. `uid` is an exception. We have:

- (name, namespace) - Set at runtime by the user or the api server, unchanged throughout execution
- (uid) - Set at runtime by the controller, unchanged throughout execution
- (results) - Set at runtime by the execution of a Task, after it completes
- (status) - Set at runtime by the controller, after Task completion (maybe one day during execution if we add dag task support)

TL;DR I think @sbwsg makes a good point about the inconsistencies, and i wonder if we might want to revisit `context` entirely - maybe we don't need that additional level? (I think I suggested it... 😢 )

ANYWAY I don't think this TEP should have to sort all of that out :D however I really like @sbwsg 's suggestion to use `(tasks.<task-name>.status)` - having to preface this with `context.pipelineRun.Tasks` seems verbose with no clear benefit? I would argue that `tasks.<taskName>.results.<resultName>` is "runtime information about the execution of a Task" and it would make sense to include status there as well?

(At the very least is there a version where `$(context.pipelineRun.Tasks.<pipelineTask>.status)` is shorter)

Apologies for re-opening this thread!

Reply:
I think suggesting having something like `context` was and is great.

In a `pipeline`, `context` helps distinguish user specified variable from the one available out of the box i.e. the pattern `context.pipelineRun.[*]` identifies that any variable with this kind of pattern will be populated with appropriate value without explicitly mapping to any pipeline param (`$param.foo`). Its more readable in this [example](https://github.com/tektoncd/pipeline/blob/2a4e707664455a8e87d770939f57b468f106da65/examples/v1beta1/pipelineruns/pipelinerun-task-execution-status.yaml).

The task result consumption `tasks.<taskName>.results.<resultName>` follows different pattern than the one task execution status `$(context.pipelineRun.tasks.<pipelineTask>.status)` and a little verbose but the clear benefit is readability and easily distinguishable from a standard task/pipeline param.

if the argument is not convincing, we could definitely change it to `tasks.<task-name>.status` 😄

Reply:
if we want to go with `$(tasks.<task-name>.status)`, finally can follow  `$(finally.<task-name>.status)` 😉

Reply:
I'm on the fence!

I think what you're saying about `context` being more like "not user provided values" make sense but again I think `tasks.<taskName>.results.<resultName>` breaks the mold in being kinda user provided (but via a task they may not have written) and kinda provided by the controller (only available as a value at runtime)

Anyway I'm happy to defer to whichever you prefer @pritidesai and we can always add support for whichever you don't choose later if we feel we need to. (I feel like it's not too crazy to end up in a state where there is more than one way to refer to the same var, b/c otherwise we might get stuck looking for the perfect val)

Reply:
thanks @bobcatfish and @sbwsg for being patient.

I have renamed the variable to `$(tasks.<taskName>.status)` and documented the reasoning [here](https://github.com/tektoncd/community/pull/234/files#diff-cdb85680663e5bee0e2b91a5bc41cb4f0fd1135f7f41aa6cf5f44d02070cc492R110). 

After staring at it for a while, started to look too verbose for its purpose 😜